### PR TITLE
WI #1539 Implement Equals/GetHashCode IEquatable<T> ruleset

### PR DESCRIPTION
**Pull request by LANOYDO**

### CODE REVIEW

File: TypeCobol/Compiler/CodeElements/CodeElement.cs [hunk 1]
```diff
@@ -169,6 +177,11 @@ public override bool Equals(object obj)
                    this.Text == codeElement.Text;
```
Comment (smedilol) [inline TypeCobol/Compiler/CodeElements/CodeElement.cs]: Do we need to compare the full text of the CodeElement ?
I think if the CodeElement contains consumedTokens, then the first token should be enough to compare.
CodeElement cannot overlaps.

File: TypeCobol/Compiler/CodeElements/CodeElement.cs [hunk 2]
```diff
@@ -169,6 +177,11 @@ public override bool Equals(object obj)
...
         }
 
+        public override int GetHashCode()
+        {
+            return _type.GetHashCode() * 11 + ConsumedTokens.GetHashCode();
```
Comment (smedilol) [inline TypeCobol/Compiler/CodeElements/CodeElement.cs]: GetHashCode doesn't compare the same properties than Equals, is it ok?
Same remarks than for Equals:
I think if the CodeElement contains consumedTokens, then the first token should be enough to compare.
CodeElement cannot overlaps.

File: TypeCobol/Compiler/CodeElements/Expressions/StorageArea.cs [hunk 3]
```diff
@@ -70,6 +70,21 @@ public override string ToString()
...
+        }
+
+        public bool Equals(StorageArea storageArea)
+        {
+            return System.Object.ReferenceEquals(this, storageArea);
+        }
+
+        public override int GetHashCode()
+        {
+            return ToString().GetHashCode();
```
Comment (smedilol) [inline TypeCobol/Compiler/CodeElements/Expressions/StorageArea.cs]: Equals use ReferenceEquals, but GetHashCode use ToString().
Why ?

File: TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs [hunk 4]
```diff
@@ -117,13 +117,16 @@ public override bool AcceptASTVisitor(IASTVisitor astVisitor)
...
+    public class SymbolReference : SymbolInformation, IEquatable<SymbolReference>
     {
+        private readonly int _hashCode;
+
         public SymbolReference(SyntaxValue<string> nameLiteral, SymbolType type) :
             base(nameLiteral, SymbolRole.SymbolReference, type)
         {
             IsAmbiguous = false;
             IsQualifiedReference = false;
+            _hashCode = nameLiteral.GetHashCode() + type.GetHashCode();
```
Comment (smedilol) [inline TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs]: Why use a cache value for _hashCode in this class and not the others ?

File: TypeCobol/Compiler/Diagnostics/Diagnostic.cs [hunk 5]
```diff
@@ -26,6 +28,8 @@ public Diagnostic(MessageCode messageCode, int columnStart, int columnEnd, int l
                 CatchedException = messageArgs.FirstOrDefault(x => x is Exception) as Exception;
                 MessageArgs = messageArgs;
             }
+
+            _hashCode = (Info==null) ? $"{Line}{ColumnStart}{ColumnEnd}{Message}".GetHashCode() : $"{Line}{ColumnStart}{ColumnEnd}{Message}{Info.Code}{Info.Severity}".GetHashCode();
```
Comment (smedilol) [inline TypeCobol/Compiler/Diagnostics/Diagnostic.cs]: There are different technique to calculate the hashCode in this PR:
Sometimes, it combines hash code of different properties and here it use a string concatenation.
What is the best practice ?

File: TypeCobol/Compiler/CodeElements/CodeElement.cs [hunk 6]
```diff
@@ -13,13 +14,16 @@ namespace TypeCobol.Compiler.CodeElements
...
+    public abstract partial class CodeElement: IToken, IVisitable, IEquatable<CodeElement>
     {
+        private readonly CodeElementType _type;
+
         protected CodeElement(CodeElementType type)
         {
             Type = type;
             ConsumedTokens = new List<Token>();
             SymbolInformationForTokens = new Dictionary<Token, SymbolInformation>();
+            _type = type;
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/CodeElement.cs]: _type is a readonly field; Type is a non-readonly property.
The call to GetHashCode needs a readonly field/property

[the remark this answers is not in the document]
Comment (smedilol) [reply TypeCobol/Compiler/CodeElements/CodeElement.cs]: Can you try to mark "Type" as readonly ?

[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/CodeElement.cs]: "Type" as readonly done.
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/77c907c8055e95bf877861ca4745d1033758792c

File: TypeCobol/Compiler/CodeElements/CodeElement.cs (hunk 1 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/CodeElement.cs]: done: the comparaison uses "Type" and "ConsumedTokens".
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/77c907c8055e95bf877861ca4745d1033758792c

File: TypeCobol/Compiler/CodeElements/CodeElement.cs (hunk 2 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/CodeElement.cs]: done: Equals and GetHashCode use "Type" and "ConsumedTokens".
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/77c907c8055e95bf877861ca4745d1033758792c

File: TypeCobol/Compiler/CodeElements/Expressions/StorageArea.cs (hunk 3 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/Expressions/StorageArea.cs]: The overrides of Equals/GetHashCode, hence IEquatable<StorageArea>, are not necessary; they have been removed.
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/77c907c8055e95bf877861ca4745d1033758792c

File: TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs (hunk 4 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs]: The reason was _hashCode is a readonly field; Type is a non-readonly property.
The call to GetHashCode needs a readonly field/property.
Modification has been made to use "Type" and "Name" changed in readonly properties.
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/77c907c8055e95bf877861ca4745d1033758792c

File: TypeCobol/Compiler/Diagnostics/Diagnostic.cs (hunk 5 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/Diagnostic.cs]: Use of fields/properties.
The overrides of Equals/GetHashCode, hence IEquatable<Diagnostic>, are not necessary; they have been removed.
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/77c907c8055e95bf877861ca4745d1033758792c

File: TypeCobol/Compiler/CodeElements/CodeElement.cs [hunk 7]
```diff
@@ -157,16 +159,20 @@ public override string ToString()
...
-                   this.Text == codeElement.Text;
+            return Equals(obj as CodeElement);
+        }
+
+        public bool Equals(CodeElement codeElement)
+        {
+            if (Object.ReferenceEquals(this, codeElement)) return true;
+            if (codeElement == null) return false;
+
+            return Type == codeElement.Type && ConsumedTokens.Equals(codeElement.ConsumedTokens);
```
Comment (smedilol) [inline TypeCobol/Compiler/CodeElements/CodeElement.cs]: I think Equals on list will only compare reference of list not their content.
As CodeElement cannot overlaps, I think we only need to compare the first token of each ConsumedTokens list.

File: TypeCobol/Compiler/CodeElements/CodeElement.cs [hunk 8]
```diff
@@ -157,16 +159,20 @@ public override string ToString()
...
+        }
+
+        public override int GetHashCode()
+        {
+            return Type.GetHashCode() * 11 + ConsumedTokens.GetHashCode();
```
Comment (smedilol) [inline TypeCobol/Compiler/CodeElements/CodeElement.cs]: Same here:
I think GetHashCode() will certainly return a hashCode based on the reference, not the content of the list.

File: Codegen/src/Codegen.csproj [hunk 9]
```diff
@@ -13,6 +13,9 @@
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\</SolutionDir>
     <RestorePackages>true</RestorePackages>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'EI_Debug|AnyCPU'">
+    <CodeAnalysisRuleSet>..\..\TypeCobol\TypeCobol.ruleset</CodeAnalysisRuleSet>
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply Codegen/src/Codegen.csproj]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/ca8920a3aabefc6b8a0d435082b58f0dbb1d4949

File: TypeCobol/Compiler/Nodes/Data.cs [hunk 10]
```diff
@@ -850,18 +850,15 @@ public override bool VisitNode(IASTVisitor astVisitor)
 
         public override bool Equals(object obj)
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/Nodes/Data.cs]: It doesn't work without overriding Equals and GetHashCode.
Equals(TypeDefinition) works only with ReferenceEquals; this method has been modified.
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/ca8920a3aabefc6b8a0d435082b58f0dbb1d4949

File: TypeCobol/Compiler/Source/StringSourceText.cs [hunk 11]
```diff
@@ -372,27 +372,32 @@ public override int Size
...
-        //public override int GetHashCode ()
-        //{
-        //    int hash;
-        //    int i;
-
-        //    for (hash = 0, i = 0; i < Size; i++)
-        //        hash= (hash << 1) ^ content[i];
-        //    return hash;
-        //}
+        public override int GetHashCode()
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/Source/StringSourceText.cs]: It's right; overrides of GetHashCode and Equals have been removed.
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/ca8920a3aabefc6b8a0d435082b58f0dbb1d4949

File: TypeCobol/Compiler/CodeElements/CodeElement.cs (hunk 7 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/CodeElement.cs]: It's right; it's modified by using ConsumedTokens[0] if it exists.
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/ca8920a3aabefc6b8a0d435082b58f0dbb1d4949

File: TypeCobol/Compiler/CodeElements/CodeElement.cs (hunk 8 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/CodeElement.cs]: It's right; it's modified by using ConsumedTokens[0] if it exists.
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/ca8920a3aabefc6b8a0d435082b58f0dbb1d4949

File: TypeCobol/Compiler/CodeElements/CodeElement.cs [hunk 12]
```diff
@@ -157,16 +159,36 @@ public override string ToString()
...
-                   this.Column == codeElement.Column &&
-                   this.StartIndex == codeElement.StartIndex &&
-                   this.StopIndex == codeElement.StopIndex &&
-                   this.Text == codeElement.Text;
+            return Equals(obj as CodeElement);
+        }
+
+        public bool Equals(CodeElement codeElement)
+        {
+            if (Object.ReferenceEquals(this, codeElement)) return true;
```
Comment (smedilol) [inline TypeCobol/Compiler/CodeElements/CodeElement.cs]: Can you try without overriding Equals and GetHashCode?

Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/CodeElement.cs]: It doesn't work without overriding Equals and GetHashCode.

File: TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs [hunk 13]
```diff
@@ -166,6 +167,31 @@ public override bool AcceptASTVisitor(IASTVisitor astVisitor)
...
+            if (Object.ReferenceEquals(this, symbolReferenceCompare)) return true;
+
+            return NameLiteral == symbolReferenceCompare.NameLiteral &&
+                   Role == symbolReferenceCompare.Role &&
+                   Type == symbolReferenceCompare.Type;
+        }
+
+        public override int GetHashCode()
+        {
+            return NameLiteral.GetHashCode() + Role.GetHashCode() + Type.GetHashCode();
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs]: It's fixed by Type.GetHashCode() * 11 + ...
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/da02539fad5306ad1f73795d818154268e07129a

File: TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs [hunk 14]
```diff
@@ -166,6 +167,31 @@ public override bool AcceptASTVisitor(IASTVisitor astVisitor)
...
+            return ComputeHashCode();
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs]: it's an error.
https://github.com/TypeCobolTeam/TypeCobol/pull/1555/commits/da02539fad5306ad1f73795d818154268e07129a

File: TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs [hunk 15]
```diff
@@ -166,6 +177,33 @@ public override bool AcceptASTVisitor(IASTVisitor astVisitor)
...
+            unchecked
+            {
+                var hashCode = 13;
+                hashCode = (hashCode * 397) ^ Type.GetHashCode();
+                hashCode = (hashCode * 397) ^ NameLiteral.GetHashCode();
+                hashCode = (hashCode * 397) ^ Role.GetHashCode();
+
+                return hashCode;
+            }
+        }
```
Comment (fm-117) [inline TypeCobol/Compiler/CodeElements/Expressions/SymbolName.cs]: SymbolReference inherits from SymbolInformation so it seems odd to use the Role here and not in the base class. Could you check whether it works or not without overriding anything in SymbolReference, just using the SymbolInformation comparison mechanism ?

File: TypeCobol/Compiler/Nodes/Data.cs [hunk 16]
```diff
@@ -843,7 +843,7 @@ public override bool VisitNode(IASTVisitor astVisitor)
         }
     }
     // [COBOL 2002]
-    public class TypeDefinition: DataDefinition, Parent<DataDescription>, IDocumentable
+    public class TypeDefinition: DataDefinition, Parent<DataDescription>, IDocumentable, IEquatable<TypeDefinition>
```
Comment (fm-117) [inline TypeCobol/Compiler/Nodes/Data.cs]: TypeDefinition and GeneratedDefinition are special cases when it comes to equality mechanism... They should be changed in a new issue. As discussed together, those two should not be modified in #1586

File: TypeCobol.LanguageServer/VsCodeProtocol/Signature Help/SignatureInformation.cs [hunk 17]
```diff
@@ -41,14 +42,32 @@ public SignatureInformation(string label, string documentation, params Parameter
...
-                   && CompareLists(this.parameters.ToList(), givenSignature.parameters.ToList());
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                var hashCode = 13;
+                hashCode = (hashCode * 397) ^ label.GetHashCode();
+                var docHashCode = documentation?.GetHashCode() ?? 10;
+                hashCode = (hashCode * 397) ^ docHashCode;
+                var paramHashCode = parameters?.GetHashCode() ?? 100;
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol.LanguageServer/VsCodeProtocol/Signature Help/SignatureInformation.cs]: done

File: TypeCobol/Compiler/CodeElements/CodeElement.cs [hunk 18]
```diff
@@ -157,16 +159,48 @@ public override string ToString()
...
+
+            if (ConsumedTokens == null && codeElement.ConsumedTokens == null)
+            {
+                return Type == codeElement.Type;
+            }
+            if (ConsumedTokens != null && codeElement.ConsumedTokens != null)
+            {
+                if (ConsumedTokens.Count > 0 && codeElement.ConsumedTokens.Count > 0)
+                {
+                    return Type == codeElement.Type && Object.ReferenceEquals(ConsumedTokens[0], codeElement.ConsumedTokens[0]);
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol/Compiler/CodeElements/CodeElement.cs]: done

**Comment by smedilol** [comment]
@LANOYDO I thin this PR is now replaced by PR #1600, right ?

**Comment by LANOYDO** [comment]
> @LANOYDO I thin this PR is now replaced by PR #1600, right ?

@LANOYDO : yes this PR is replaced by PR #1600

**Comment by rooksdo** [comment]
PR replaced by #1600

**Pull request closed without merging**